lib/services/local/plugin_data.go

```diff
@@ -0,0 +1,219 @@
+package local
+
+import (
+	"bytes"
+	"context"
+	"time"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+)
+
+// PluginDataService manages PluginData
+type PluginDataService struct {
+	backend.Backend
+}
+
+// NewPluginDataService creates a new PluginDataService
+func NewPluginDataService(backend backend.Backend) *PluginDataService {
+	return &PluginDataService{backend}
+}
+
+// GetPluginData loads all plugin data matching the supplied filter.
+func (s *PluginDataService) GetPluginData(ctx context.Context, filter types.PluginDataFilter) ([]types.PluginData, error) {
+	data, err := s.getPluginData(ctx, filter)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return data, nil
+}
+
+// getPluginData returns an array of PluginData matching filter criteria
+func (s *PluginDataService) getPluginData(ctx context.Context, filter types.PluginDataFilter) ([]types.PluginData, error) {
+	// Filters which specify Resource are a special case since they will match exactly zero or one
```

Comment:
Two changes: 
  1) No need to separate it into multiple functions since all resource kinds (for now) use the same logic.
  2) Let's move the filter kind check back to here and out of the server.
```suggestion
// GetPluginData loads all plugin data matching the supplied filter.
func (s *PluginDataService) GetPluginData(ctx context.Context, filter types.PluginDataFilter) ([]types.PluginData, error) {
	if !isAllowedPluginDataType(filter.Kind) {
		return nil, trace.BadParameter("unsupported resource kind %q", filter.Kind)
	}
	
	// Filters which specify Resource are a special case since they will match exactly zero or one
```